src/main/java/org/broadinstitute/hellbender/tools/walkers/filters/VariantFiltration.java

```diff
@@ -0,0 +1,389 @@
+                if (filterRecordsNotInMask) {
+                    hInfo.add(new VCFFilterHeaderLine(maskName, "Doesn't overlap a user-input mask"));
+                } else {
+                    hInfo.add(new VCFFilterHeaderLine(maskName, "Overlaps a user-input mask"));
+                }
+            }
+        } catch (final IllegalArgumentException e) {
+            throw new UserException.BadInput(e.getMessage());
+        }
+
+        writer.writeHeader(new VCFHeader(hInfo, getHeaderForVariants().getGenotypeSamples()));
+    }
+
+    @Override
+    public void onTraversalStart() {
+        if ( maskExtension < 0 ) {
+            throw new UserException.BadArgumentValue("maskExtension", "negative values are not allowed");
+        }
+
+        if (filterRecordsNotInMask && mask == null) {
+            throw new UserException.BadArgumentValue("filterNotInMask", "argument not allowed if mask argument is not provided");
+        }
+        filterExps = VariantContextUtils.initializeMatchExps(filterNames, filterExpressions);
+        genotypeFilterExps = VariantContextUtils.initializeMatchExps(genotypeFilterNames, genotypeFilterExpressions);
+
+        VariantContextUtils.engine.get().setSilent(true);
+
+        initializeVcfWriter();
+    }
+
+    @Override
+    public void apply(final VariantContext variant, final ReadsContext readsContext, final ReferenceContext ref, final FeatureContext featureContext) {
+
+        final VariantContext vc1 = invalidatePreviousFilters ? (new VariantContextBuilder(variant)).filters(Collections.emptySet()).make() : variant;
+        final VariantContext vc = filterPresent(vc1) ? vc1: addMaskIfCoversVariant(vc1, featureContext);
+
+        filter(vc, featureContext);
+    }
+
+    /**
+     * Add mask to variant context filters if it covers the it's location
+     * @return VariantContext with the mask added if the VariantContext is within the extended mask area
+     */
+    private VariantContext addMaskIfCoversVariant(final VariantContext vc, final FeatureContext featureContext) {
+        final List<Feature> maskSnps = featureContext.getValues(mask, maskExtension, maskExtension);
+
+        final boolean snpsMasked = maskSnps.isEmpty() == filterRecordsNotInMask;
+        if (snpsMasked) {
+            final Set<String> oldFiltersPlusNewOne = Sets.union(vc.getFilters(), singleton(maskName));
+            return new VariantContextBuilder(vc).filters(oldFiltersPlusNewOne).make();
+        } else {
+            return vc;
+        }
+    }
+
+    private boolean filterPresent(final VariantContext vc) {
+        return vc.getFilters() != null && vc.getFilters().contains(maskName);
+    }
+
+    private void filter(final VariantContext vc, final FeatureContext featureContext) {
+        final VariantContextBuilder builder = new VariantContextBuilder(vc);
+
+        // make new Genotypes based on filters
+        if ( !genotypeFilterExps.isEmpty() || setFilteredGenotypesToNocall ) {
```

Comment:
yes, i think so. There may already be filters in the genotypes and then the set them to no call if `setFilteredGenotypesToNocall` is true